mikeio/dfsu.py

```diff
@@ -488,6 +499,119 @@ def _get_element_coords(self):
+        except:
+            warnings.warn('Could not determine if domain contains points. Failed to convert to_shapely()')
+            return None
+        
+        cnts = mp.Path(domain.exterior).contains_points(points)
+        for interj in domain.interiors:
+            in_hole = mp.Path(interj).contains_points(points) 
+            cnts = np.logical_and(cnts, ~in_hole)
+        return cnts
+
+
+    def _create_tree2d(self):
+        xy = self.geometry2d.element_coordinates[:,:2]
+        self._tree2d = cKDTree(xy)
+
+    def _find_n_nearest_2d_elements(self, x, y=None, n_points=1):        
+        if self._tree2d is None:
+            self._create_tree2d()
+
+        if y is None: 
+            p = x
+        else:
+            p = np.array((x, y))
+        d, elem_id = self._tree2d.query(p, k=n_points)
+        return elem_id, d
+
+    def get_overset_grid(self, dx=None, shape=None):
+        coords = self.node_coordinates   # node_ or element_
+        small = 1e-10        
+        x0 = coords[:,0].min() + small
+        y0 = coords[:,1].min() + small
+        x1 = coords[:,0].max() - small
+        y1 = coords[:,1].max() - small
+        xr = x1 - x0
+        yr = y1 - y0
+        
+        if (dx is None) and (shape is None):
+            if xr <= yr:
+                nx = 10
+                ny = int(np.ceil(yr/xr))*nx
+            else:
+                ny = 10
+                nx = int(np.ceil(xr/yr))*ny
+            dx = xr/(nx-1)
+            dy = yr/(ny-1)
+        else:
+            if dx is None:
+                nx, ny = shape
+                dx = xr/(nx-1)
+                dy = yr/(ny-1)
+            if shape is None:
+                dy = dx
+                nx = int(np.ceil(xr/dx)) + 1
+                ny = int(np.ceil(yr/dy)) + 1
+
+        return x0, dx, nx, y0, dy, ny
+
+    def get_2d_interpolant(self, xy, n_nearest:int=1):
+        elem_ids, distances = self._find_n_nearest_2d_elements(xy, n_points=n_nearest)
+
+        if n_nearest == 1:
+            weights = np.ones(distances.shape)
+            weights[~self.contains(xy)] = np.nan
+        elif n_nearest > 1:
+            p = 1   # inverse distance order 
+            weights = 1 / distances**p
+            denom = weights.sum(axis=1).reshape(-1,1)*np.ones((1,n_nearest))
+            weights = weights / denom
+            weights[~self.contains(xy),:] = np.nan
+        else:
+            ValueError('n_nearest must be at least 1')
+
+        return elem_ids, weights
+
+    def gridded_2d_interpolant(self, x0, dx, nx:int, y0, dy, ny:int, n_nearest:int=1):
+        x1 = x0 + dx*(nx-1)
+        x = np.linspace(x0, x1, nx)
+        y1 = y0 + dy*(ny-1)
+        y = np.linspace(y0, y1, ny)
+        xv, yv = np.meshgrid(x, y)
+        xy = np.empty((nx*ny, 2))
+        xy[:,0] = xv.reshape(-1)
+        xy[:,1] = yv.reshape(-1)
+        return self.get_2d_interpolant(xy, n_nearest)
+
+    def interp2d(self, data, interpolant2d):
```

Comment:
interp2d could be extracted into a new class

Reply:
good idea

Reply:
I have now extracted functionality into a new file - but could not really see the benefit of making it a class - maybe you can?

Reply:
No.

Reply:
Okay - but I'll guess we will keep it in a separate file for now, right?